### PR TITLE
Job Endpoints (Create, Read, Update, Delete) & Validation Schema

### LINKED ISSUE #26: Create, Read Job
See [Database Schema Doc](https://github.com/uoftblueprint/tces/wiki/Database-Schema) for info.

API Endpoint info:
- GET `/api/job/{job_id}/`. No headers, response should be JSON payload of job with id `job_id`. 404 if not found.
- POST `/api/job`. Expects to receive all the fields for a new `job` object in the request body.

This should be split up into:
* a class in Model directory (Handle interacting with MongoDB directly)
* a class in Controller directory (This will use your class in Model directory to perform the expected functionality for each endpoint)
* a class in Routes directory (This will use functions from the class you define in the Controller directory, but actually tie them to the endpoint)

See [project structure
](https://github.com/uoftblueprint/tces/wiki/Project-Structure) for further reference. Also msg me for clarification, I think this is a pretty confusing concept.

### PR DESCRIPTION
**Pull request by helen-li**
# Description

Fixes #26, #49  

* Created model, controller, and route for job endpoints (GET, POST, PUT, DELETE)
* Added jobs schema in `conn.ts` and connected it to the jobs collection
* Dependency added in server dir for testing: `@jest-mock/express": "^2.0.1`
* Also ignore the 2 `client` dir files in this PR, I didn't change them but Git argues otherwise lol

## Sample curl for GET
curl --location --request GET 'http://localhost:3001/api/job/63aa4054c792bf4a6580bfd4/'

## Response for GET
{
    "_id": "63aa4054c792bf4a6580bfd4",
    "title": "Test Job XII",
    "creation_date": "2022-07-21T00:00:00.000Z",
    "creator_id": "63bb1f16939f28affbacf5fa",
    "owner_id": "63bb1f16939f28affbacf5fa",
    "timeline": {},
    "employer_id": "63aa1f11cb4203e32ed5264b",
    "pay_per_hour": 22.5,
    "address_id": null,
    "description": null,
    "type": null,
    "expiry_date": null
}

## Sample curl for POST
curl --location --request POST 'http://localhost:3001/api/job' \
--header 'Content-Type: application/json' \
--data-raw '{
    "title": "Test Job XII",
    "creation_date": "2022-07-21",
    "creator_id": "63bb1f16939f28affbacf5fa",
    "timeline": {},
    "employer_id": "63aa1f11cb4203e32ed5264b",
    "pay_per_hour": 22.5
}'

## Response for POST
Successfully created a new job with id 63aa4054c792bf4a6580bfd4 (should be a different ID if you try POST in Postman)

## Sample curl for DELETE
curl --location --request DELETE 'http://localhost:3001/api/job/63aa4054c792bf4a6580bfd4/'

## Response for DELETE
Successfully removed job with id 63aa4054c792bf4a6580bfd4 (whatever ID is in the call)

## Sample curl for PUT
curl --location --request PUT 'http://localhost:3001/api/job/63aa4054c792bf4a6580bfd4/' \
--header 'Content-Type: application/json' \
--data-raw '{
    "title": "Test Updating Job",
    "creation_date": "2022-12-09",
    "creator_id": "63aa1f11cb4203e32ed5264b",
    "owner_id": "63aa40e680f1d80909741d03",
    "timeline": {},
    "employer_id": "63aa1f11cb4203e32ed5264b"
}'

## Response for PUT
Successfully updated job with id 63aa4054c792bf4a6580bfd4 (whatever ID is in the call)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] I have run `npm run lint:fix`
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if required)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (Delete if frontend)

### CODE REVIEW

File: server/controllers/job.controller.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
+// External Dependencies
+import { NextFunction, Request, Response } from "express";
+import Job from "../models/job.model";
+
+const getJobById = async (req: Request, res: Response, next: NextFunction) => {
+    try {
+        const id = req?.params?.id;
```
Comment (EclipseIsDead) [inline server/controllers/job.controller.ts:7]: I think these are unsafe, might be a better TS way of doing it.

File: server/controllers/job.controller.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        const job = await Job.getJobById(id);
+        res.status(200).json(job);
+    } catch (error) {
+        if (error instanceof Error) {
+            console.log(error);
```
Comment (EclipseIsDead) [inline server/controllers/job.controller.ts:12]: logger

File: server/src/database/conn.ts [hunk 3]
```diff
@@ -26,8 +25,71 @@ export async function connectToDatabase() {
...
 
-
-
+  await db.command({
+    "collMod": JOBS_COLLECTION_NAME,
+    "validator": jobSchema
+  });
 }
 
+const jobSchema = {
```
Comment (EclipseIsDead) [inline server/src/database/conn.ts:40]: ideally this is a separate file

File: server/test/unit/job.test.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+const mockCreateJobError = new Error("Failed to create a new job.");
+const mockDeleteJobError = new Error(`Failed to remove job with id ${mockErrorJobId}`);
+const mockUpdateJobError = new Error(`Job with id: ${mockErrorJobId} not updated`);
+
+describe('Job Controller Unit Tests', () => {
+  beforeEach(() => {
+    clearMockRes();
+  });
+
+  test('should get job by id successfully', async () => {
```
Comment (EclipseIsDead) [inline server/test/unit/job.test.ts:54]: use the logger here too for confirmation

**Review by EclipseIsDead — CHANGES_REQUESTED** [review]
really good stuff, small changes

File: server/test/unit/job.test.ts (hunk 4 above)
Comment (helen-li) [reply server/test/unit/job.test.ts:57]: i just added some info logs at the bottom of each test, lmk if this is not what you intended to suggest lol

File: server/controllers/job.controller.ts (hunk 1 above)
Comment (helen-li) [reply server/controllers/job.controller.ts:10]: i just followed mongo's guide for typescript integration & this is the exact code they had lol - lemme dig around a bit more to check for maybe better alternatives

Comment (EclipseIsDead) [reply server/controllers/job.controller.ts:10]: No worries, double checked the params its all g

**Review by EclipseIsDead — APPROVED** [review]
lgtm

**Review by kenneth-miura — CHANGES_REQUESTED** [review]
can u move all the backend directories to be inside of src?

File: server/src/database/conn.ts [hunk 5]
```diff
@@ -26,8 +25,14 @@ export async function connectToDatabase() {
...
+  collections.jobs = jobsCollection;
+  console.log(
+    `Successfully connected to collection: ${jobsCollection.collectionName}`
+  );
 
-
-
-}
-
+  await db.command({
```
Comment (jordanjanakievski) [inline server/src/database/conn.ts:34]: `await db.createCollection(JOBS_COLLECTION_NAME, {
		validator: jobSchema,
	});`
Adding this code in the creation of the collection will validate and and result in a successful connection to the backend

**Review by jordanjanakievski — COMMENTED** [review]
Added a comment in conn.ts that should fix the validation issue

**Review by kenneth-miura — APPROVED** [review]
image

^u

**Review by jordanjanakievski — APPROVED** [review]
Backend connected and requests were successful. Validation looked good.

**Pull request merged**